File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/CategoryPagerAdapter.java
package ch.rmy.android.http_shortcuts;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import java.util.ArrayList;
import java.util.List;

import ch.rmy.android.http_shortcuts.realm.models.Category;

public class CategoryPagerAdapter extends FragmentPagerAdapter {

    private final FragmentManager fragmentManager;

    private List<ListFragment> fragments = new ArrayList<>();

    public CategoryPagerAdapter(FragmentManager fragmentManager) {
        super(fragmentManager);
        this.fragmentManager = fragmentManager;
    }

    public void setCategories(List<Category> categories, boolean shortcutPlacementMode) {
        for (int i = fragments.size(); i < categories.size(); i++) {
            Fragment restoredFragment = fragmentManager.findFragmentByTag(makeFragmentName(i));
            if (restoredFragment != null && restoredFragment instanceof ListFragment) {
                fragments.add((ListFragment) restoredFragment);
            } else {
                fragments.add(new ListFragment());
            }
        }

        for (int i = 0; i < fragments.size(); i++) {
            ListFragment fragment = fragments.get(i);
            fragment.setCategoryId(categories.get(i).getId());
            fragment.setShortcutPlacementMode(shortcutPlacementMode);
        }
    }

    @Override
    public ListFragment getItem(int position) {
        return fragments.get(position);
    }

    @Override
    public int getCount() {
        return fragments.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return fragments.get(position).getCategoryName();
    }

    private static String makeFragmentName(int position) {
        return "android:switcher:" + R.id.view_pager + ":" + position;
    }

}

File: HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/activities/main/CategoryPagerAdapter.kt
package ch.rmy.android.http_shortcuts.activities.main

import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.data.enums.SelectionMode

class CategoryPagerAdapter(
    private val fragmentManager: FragmentManager,
) : FragmentPagerAdapter(
    fragmentManager,
    BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT
) {

    private var fragments: List<Pair<String, ShortcutListFragment>> = emptyList()

    private var previousCategories: List<CategoryTabItem>? = null

    fun setCategories(categories: List<CategoryTabItem>, selectionMode: SelectionMode) {
        if (categories == previousCategories) {
            return
        }
        previousCategories = categories
        fragments = categories
            .mapIndexed { index, category ->
                val fragment = fragmentManager.findFragmentByTag(makeFragmentName(index))
                    ?.let { it as? ShortcutListFragment }
                    ?.takeIf { it.categoryId == category.categoryId && it.layoutType == category.layoutType && it.selectionMode == selectionMode }
                    ?: ShortcutListFragment.create(category.categoryId, category.layoutType, selectionMode)
                category.name to fragment
            }
        notifyDataSetChanged()
    }

    override fun getItem(position: Int) = fragments[position].second

    override fun getCount() = fragments.size

    override fun getPageTitle(position: Int) = fragments[position].first

    private fun makeFragmentName(position: Int) = "android:switcher:" + R.id.view_pager + ":" + position
}
